<?xml version='1.0' encoding='UTF-8'?>
<root>
 <tabbed_widget parent="main_window" name="Main Window">
  <Tab containers="1" tab_name="Forces&amp;Positions">
   <Container>
    <DockSplitter orientation="-" sizes="0.501832;0.498168" count="2">
     <DockArea name="...">
      <plot flip_x="false" flip_y="false" style="Lines" mode="TimeSeries">
       <range left="2.019970" bottom="-0.001000" top="1.200000" right="37.009983"/>
       <limitY max="1.2" min="-0.001"/>
       <curve name="/myrmex_controller_debug/f.0" color="#1f77b4"/>
       <curve name="/myrmex_controller_debug/f.1" color="#d62728"/>
       <curve name="/myrmex_controller_debug/f_thresholds.0" color="#1ac938"/>
       <curve name="/myrmex_controller_debug/f_target" color="#1f77b4"/>
       <curve name="/myrmex_controller_debug/f_sum" color="#ff7f0e"/>
      </plot>
     </DockArea>
     <DockSplitter orientation="|" sizes="0.502638;0.497362" count="2">
      <DockArea name="...">
       <plot flip_x="false" flip_y="false" style="Lines" mode="TimeSeries">
        <range left="2.019970" bottom="-0.010000" top="0.050000" right="37.009983"/>
        <limitY max="0.05" min="-0.01"/>
        <curve name="/myrmex_controller_debug/des_q.0" color="#ff7f0e"/>
        <curve name="/myrmex_controller_debug/des_q.1" color="#f14cc1"/>
        <curve name="/myrmex_gripper_controller/state/actual/positions.0" color="#9467bd"/>
        <curve name="/myrmex_gripper_controller/state/actual/positions.1" color="#17becf"/>
       </plot>
      </DockArea>
      <DockSplitter orientation="-" sizes="0.501229;0.498771" count="2">
       <DockArea name="...">
        <plot flip_x="false" flip_y="false" style="Lines" mode="TimeSeries">
         <range left="2.019970" bottom="-0.000786" top="0.051142" right="37.009983"/>
         <limitY/>
         <curve name="/myrmex_controller_debug/delta_q" color="#1f77b4"/>
        </plot>
       </DockArea>
       <DockArea name="...">
        <plot flip_x="false" flip_y="false" style="Lines" mode="TimeSeries">
         <range left="2.019970" bottom="-0.000068" top="0.002799" right="37.009983"/>
         <limitY/>
         <curve name="/myrmex_controller_debug/error_integral" color="#d62728"/>
        </plot>
       </DockArea>
      </DockSplitter>
     </DockSplitter>
    </DockSplitter>
   </Container>
  </Tab>
  <currentTabIndex index="0"/>
 </tabbed_widget>
 <use_relative_time_offset enabled="1"/>
 <!-- - - - - - - - - - - - - - - -->
 <!-- - - - - - - - - - - - - - - -->
 <Plugins>
  <plugin ID="DataLoad CSV">
   <default time_axis="" delimiter="0"/>
  </plugin>
  <plugin ID="DataLoad ROS bags">
   <use_header_stamp value="false"/>
   <discard_large_arrays value="true"/>
   <max_array_size value="100"/>
   <boolean_strings_to_number value="true"/>
   <remove_suffix_from_strings value="true"/>
  </plugin>
  <plugin ID="DataLoad ULog"/>
  <plugin ID="ROS Topic Subscriber">
   <use_header_stamp value="true"/>
   <discard_large_arrays value="true"/>
   <max_array_size value="100"/>
   <boolean_strings_to_number value="true"/>
   <remove_suffix_from_strings value="true"/>
  </plugin>
  <plugin ID="UDP Server"/>
  <plugin ID="WebSocket Server"/>
  <plugin ID="Fast Fourier Transform"/>
  <plugin ID="Quaternion to RPY"/>
  <plugin ID="Reactive Script Editor">
   <library code="--[[ Helper function to create a series from arrays&#xa;&#xa; new_series: a series previously created with ScatterXY.new(name)&#xa; prefix:     prefix of the timeseries, before the index of the array&#xa; suffix_X:   suffix to complete the name of the series containing the X value. If [nil], use the index of the array.&#xa; suffix_Y:   suffix to complete the name of the series containing the Y value&#xa; timestamp:   usually the tracker_time variable&#xa;              &#xa; Example:&#xa; &#xa; Assuming we have multiple series in the form:&#xa; &#xa;   /trajectory/node.{X}/position/x&#xa;   /trajectory/node.{X}/position/y&#xa;   &#xa; where {N} is the index of the array (integer). We can create a reactive series from the array with:&#xa; &#xa;   new_series = ScatterXY.new(&quot;my_trajectory&quot;) &#xa;   CreateSeriesFromArray( new_series, &quot;/trajectory/node&quot;, &quot;position/x&quot;, &quot;position/y&quot;, tracker_time );&#xa;]]--&#xa;&#xa;function CreateSeriesFromArray( new_series, prefix, suffix_X, suffix_Y, timestamp )&#xa;  &#xa;  --- clear previous values&#xa;  new_series:clear()&#xa;  &#xa;  --- Append points to new_series&#xa;  index = 0&#xa;  while(true) do&#xa;&#xa;    x = index;&#xa;    -- if not nil, get the X coordinate from a series&#xa;    if suffix_X ~= nil then &#xa;      series_x = TimeseriesView.find( string.format( &quot;%s.%d/%s&quot;, prefix, index, suffix_X) )&#xa;      if series_x == nil then break end&#xa;      x = series_x:atTime(timestamp)&#x9; &#xa;    end&#xa;    &#xa;    series_y = TimeseriesView.find( string.format( &quot;%s.%d/%s&quot;, prefix, index, suffix_Y) )&#xa;    if series_y == nil then break end &#xa;    y = series_y:atTime(timestamp)&#xa;    &#xa;    new_series:push_back(x,y)&#xa;    index = index+1&#xa;  end&#xa;end&#xa;&#xa;--[[ Similar to the built-in function GetSeriesNames(), but select only the names with a give prefix. ]]--&#xa;&#xa;function GetSeriesNamesByPrefix(prefix)&#xa;  -- GetSeriesNames(9 is a built-in function&#xa;  all_names = GetSeriesNames()&#xa;  filtered_names = {}&#xa;  for i, name in ipairs(all_names)  do&#xa;    -- check the prefix&#xa;    if name:find(prefix, 1, #prefix) then&#xa;      table.insert(filtered_names, name);&#xa;    end&#xa;  end&#xa;  return filtered_names&#xa;end&#xa;&#xa;--[[ Modify an existing series, applying offsets to all their X and Y values&#xa;&#xa; series: an existing timeseries, obtained with TimeseriesView.find(name)&#xa; delta_x: offset to apply to each x value&#xa; delta_y: offset to apply to each y value  &#xa;&#xa;]]--&#xa;&#xa;function ApplyOffsetInPlace(series, delta_x, delta_y)&#xa;  -- use C++ indeces, not Lua indeces&#xa;  for index=0, series:size()-1 do&#xa;    x,y = series:at(index)&#xa;    series:set(index, x + delta_x, y + delta_y)&#xa;  end&#xa;end&#xa;"/>
   <scripts/>
  </plugin>
  <plugin ID="CSV Exporter"/>
  <plugin ID="ROS /rosout Visualization"/>
  <plugin ID="ROS Topic Re-Publisher"/>
 </Plugins>
 <!-- - - - - - - - - - - - - - - -->
 <previouslyLoaded_Datafiles/>
 <previouslyLoaded_Streamer name="ROS Topic Subscriber"/>
 <!-- - - - - - - - - - - - - - - -->
 <customMathEquations/>
 <snippets>
  <snippet name="1st_derivative">
   <global>var prevX = 0
var prevY = 0</global>
   <function></function>
   <linked_source></linked_source>
  </snippet>
  <snippet name="1st_order_lowpass">
   <global>var prevY = 0
var alpha = 0.1</global>
   <function></function>
   <linked_source></linked_source>
  </snippet>
  <snippet name="2D_velocity">
   <global>var prev_x = 0
var prev_y = 0
var prev_t = 0</global>
   <function></function>
   <linked_source></linked_source>
  </snippet>
  <snippet name="average_two_curves">
   <global></global>
   <function></function>
   <linked_source></linked_source>
  </snippet>
  <snippet name="integral">
   <global>var integral = 0</global>
   <function></function>
   <linked_source></linked_source>
  </snippet>
  <snippet name="rad_to_deg">
   <global></global>
   <function></function>
   <linked_source></linked_source>
  </snippet>
  <snippet name="remove_offset">
   <global>var is_first = true
var first_value = 0</global>
   <function></function>
   <linked_source></linked_source>
  </snippet>
  <snippet name="yaw_from_quaternion">
   <global>// source: https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles

function quaternionToYaw(x, y, z, w)
{
  // yaw (z-axis rotation)
  t1 = 2.0 * (w * z + x * y);
  t2 = 1.0 - 2.0 * (y * y + z * z);
  yaw = Math.atan2(t1, t2);

  return yaw
}</global>
   <function></function>
   <linked_source></linked_source>
  </snippet>
 </snippets>
 <!-- - - - - - - - - - - - - - - -->
</root>

